### PR TITLE
[ENH]: start refactoring pipelines module

### PR DESCRIPTION
**Pull request by dengemann**
Do not merge.
This is still broken but you can see already a bit where it's going.
Addresses multiple issues.

### CODE REVIEW

**Comment by dengemann** [comment]
I am still having some difficulty debugging the `test_pipelines.py` @DavidSabbagh @agramfort 
I absolutely want things to be indexible by column names, not integer indices.
The ExpandFeatures class must be probably updated to return a DatafFrame.
The code inside ExpandFeatures is the problem, all passes when not expanding interaction effects.
As it is in this PR, scikit-learn complains becaus –– I guess –– it cannot drop the indicator colum based on a name, as ExpanFeagures returns an array, not a DataFrame.
Will push/investigate more tomorrow.

**Comment by DavidSabbagh** [comment]
This is exactly the problem I encountered and why I gave up and finally
indexed with integer indices. I did not wanted to loose time and make
things work.
But i agree it would be much cleaner

**Comment by dengemann** [comment]
I think we have to make clear design choices here and require that the code here only supports DataFrames all the way down. It will simplify things; working on a fix in the meantime.

**Comment by dengemann** [comment]
@agramfort @DavidSabbagh there is more to be done (especially tests) but I think this is already something. If we can do the necessary to merge this, I think, we make a step ahead.

File: meegpowreg/pipelines.py [hunk 1]
```diff
@@ -17,126 +17,147 @@
...
+def make_filter_bank_model(names, method='riemann', projection_params=None,
+                           vectorization_params=None,
+                           categorical_interaction=None, scaling=None,
+                           estimator=None):
+    """Generate pipeline for filterbank models.
+
+    Compute filter bank models as used in [1]_. These models take as input
+    sensor-space covariance matrices computed from M/EEG signals in different
+    frequency bands. Then transformations are applied that can improve the
+    applicability of linear regression techniques by removing field spread.
```
Comment (agramfort) [inline meegpowreg/pipelines.py:32]:
```suggestion
    frequency bands. Then transformations are applied to improve the
    applicability of linear regression techniques by reducing the impact of
    field spread.
```

File: meegpowreg/pipelines.py [hunk 2]
```diff
@@ -17,126 +17,147 @@
...
+
+    In terms of implementation, this involves 1) projection
+    (e.g. spatial filters) and 2) vectorization (e.g. taking the log on the
+    diagonal).
+
+    .. note::
+        The resulting model expects as inputs data frames in which different
+        covarances (e.g. for different frequencies) are stored inside columns
+        indexed by ``names''.
```
Comment (agramfort) [inline meegpowreg/pipelines.py:41]:
```suggestion
        indexed by ``names``.
```

File: meegpowreg/pipelines.py [hunk 3]
```diff
@@ -17,126 +17,147 @@
...
+        we recommend comparing `riemann', `spoc' and `diag' as a baseline.
+
+    Parameters
+    ----------
+    names : list of str
+        The column names of the data frame corresponding to different
+        covariances.
+    method : str
+        The method used for extracting features from covariances. Defaults
+        to `riemann'.
```
Comment (agramfort) [inline meegpowreg/pipelines.py:60]:
```suggestion
        to ``'riemann'``.
```

File: meegpowreg/pipelines.py [hunk 4]
```diff
@@ -17,126 +17,147 @@
...
+    categorical_interaction : str
+        The column in the input data frame containing a binary descriptor
+        used to fit 2-way interaction effects.
+    scaling : scikit-learn Transformer object | None
+        Method for re-rescaling the features. Defaults to None. If None,
+        StandardScaler is used.
+    estimator : scikit-learn Estimator object.
+        The estimator object. Defaults to None. If None, RidgeCV is used with
+        generalized cross validation for the regularization parameter alpha.
+        A logarithmic space between -3 and 5 is visited (100 values).
```
Comment (agramfort) [inline meegpowreg/pipelines.py:74]: I am wondering if we should not make 3 functions

make_filter_bank_transformer
make_filter_bank_regressor
make_filter_bank_classifier

where make_filter_bank_transformer returns the first steps of the pipeline without the scaler and the RIdgeCV

Comment (dengemann) [reply meegpowreg/pipelines.py:75]: I was thinking in that direction too. `make_filter_bank_regressor/classifier` would then just be high-level convenience functions that work API-wise like the current `make_filter_bank_model`?

**Comment by agramfort** [comment]
yes !

>

**Comment by dengemann** [comment]
@agramfort next milestone unlocked :)

File: meegpowreg/pipelines.py [hunk 5]
```diff
@@ -17,126 +17,300 @@
...
+    vectorization_params : dict
+        The parameters for the vectorization step.
+    categorical_interaction : str
+        The column in the input data frame containing a binary descriptor
+        used to fit 2-way interaction effects.
+    scaling : scikit-learn Transformer object | None
+        Method for re-rescaling the features. Defaults to None. If None,
+        StandardScaler is used.
+    estimator : scikit-learn Estimator object.
+        The estimator object. Defaults to None. If None, LogisticRegression
```
Comment (agramfort) [inline meegpowreg/pipelines.py:198]:
```suggestion
        The estimator object. Defaults to None. If None, RidgeCV
```

File: meegpowreg/pipelines.py [hunk 6]
```diff
@@ -17,126 +17,300 @@
...
-          for col in range(len(fb_cols))],
-        ('drop', -1)
+
+    scaling_ = scaling
+    if scaling_ is None:
+        scaling_ = StandardScaler()
+
+    estimator_ = estimator
+    if estimator_ is None:
+        estimator_ = LogisticRegression()
```
Comment (agramfort) [inline meegpowreg/pipelines.py:308]:
```suggestion
        estimator_ = LogisticRegression(solver='liblinear')
```
always faster from my experience

**Review by agramfort — APPROVED** [review]
besides LGTM

**Review by agramfort — APPROVED** [review]
@dengemann merge if happy

**Review by DavidSabbagh — APPROVED** [review]

**Pull request merged**